src/App.tsx

```diff
@@ -1,61 +1,147 @@
+import React, { useState } from 'react';
 import './App.scss';
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import { TodoList } from './components/TodoList';
+import { Todo } from './types/Todo';
+
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+
+const titlePattern = /^[a-zA-Zа-яА-ЯїЇіІ0-9\s]*$/; //pattern that matches 'ua' and 'en' letters, digits and spaces
+
+function getUserByUserId(userId: number) {
+  return usersFromServer.find(user => user.id === userId) || null;
+}
+
+function getPreparedTodos() {
+  const preparedTodos = todosFromServer.map(todo => ({
+    ...todo,
+    user: getUserByUserId(todo.userId),
+  }));
+
+  return preparedTodos;
+}
+
+export const App: React.FC = () => {
+  const todos = getPreparedTodos();
+
+  const [visibleTodos, setVisibleTodos] = useState(todos);
+  const [title, setTitle] = useState('');
+  const [userId, setUserId] = useState(0);
+  const [titleError, setTitleError] = useState('');
+  const [userIdError, setUserIdError] = useState('');
+
+  function getLargestTodoId() {
+    const sortedById = [...visibleTodos].sort(
+      (todo1, todo2) => todo2.id - todo1.id,
+    );
+
+    return sortedById[0].id + 1;
+  }
+
+  function titleValidation() {
+    return title
+      .split('')
+      .filter(ch => ch.match(titlePattern))
+      .join('');
+  }
```

Comment:
You can validate the whole string at once https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions